### ISSUE TITLE
lxc list hangs on cluster

### ISSUE THREAD

**Issue by chris-sanders** [lxc/lxd#7422]
# Required information
 lxc info
config:
  core.https_address: 192.168.0.103:8443
  core.proxy_ignore_hosts: 127.0.0.1,192.168.0.162,::1,localhost
  core.trust_password: true
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- candid_authentication
- candid_config
- candid_config_key
- usb_optional_vendorid
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses:
  - 192.168.0.103:8443
  architectures:
  - x86_64
  - i686
  certificate: REDACTED
  certificate_fingerprint: 95f620b41e60d37e2c1880afe678241b76cac98ca034e32aa769cb36cd3f9e79
  driver: lxc
  driver_version: 3.0.3
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 5.3.0-45-generic
  server: lxd
  server_pid: 977
  server_version: 3.0.3
  storage: ceph
  storage_version: ceph version 12.2.12 (1436006594665279fe734b4c15d7e08c13ebd777)
    luminous (stable)
  server_clustered: true
  server_name: juju-89bcd1-16-lxd-2
  project: ""

uname -r
5.3.0-45-generic

lsb_release -a
Distributor ID: Ubuntu
Description:    Ubuntu 18.04.3 LTS
Release:        18.04
Codename:       bionic

# Issue description
lxc list command hangs on the cluster, containers are currently running but I can't interact with them.
lxc cluster list works

> # lxc cluster list
> +----------------------+----------------------------+----------+--------+-------------------+
> |         NAME         |            URL             | DATABASE | STATE  |      MESSAGE      |
> +----------------------+----------------------------+----------+--------+-------------------+
> | juju-89bcd1-10-lxd-1 | https://192.168.0.113:8443 | YES      | ONLINE | fully operational |
> +----------------------+----------------------------+----------+--------+-------------------+
> | juju-89bcd1-16-lxd-2 | https://192.168.0.103:8443 | NO       | ONLINE | fully operational |
> +----------------------+----------------------------+----------+--------+-------------------+
> | juju-89bcd1-7-lxd-2  | https://192.168.0.107:8443 | YES      | ONLINE | fully operational |
> +----------------------+----------------------------+----------+--------+-------------------+
> | juju-89bcd1-8-lxd-2  | https://192.168.0.123:8443 | YES      | ONLINE | fully operational |
> +----------------------+----------------------------+----------+--------+-------------------+
> 

# Steps to reproduce
Not sure how to reproduce, I've just been using clustered LXD and have found recently that I while the containers are running I can't access them.

I do notice that ceph is listed as version 12.2 and this cluster has been upgraded to 13.2 though I don't see that as a problem.

# Information to attach
 - No dmesg output during the run or for the last 24-hours
 - running with debug this is the output where it hangs: https://pastebin.ubuntu.com/p/62xwDqWZBf/
 - Main daemon log: https://pastebin.ubuntu.com/p/XzBHvc25KF/
 - [lxc-monitor.txt](https://github.com/lxc/lxd/files/4677670/lxc-monitor.txt)
 - [strace.txt](https://github.com/lxc/lxd/files/4677745/strace.txt) 

While I haven't tried rebooting the cluster from scratch, I'm a bit concerned about doing that with out the ability to interact with lxd. I can't loop over the containers to shut them down, I'm concerned about them not coming up at all on a reboot, and even if that resolves the current issue I don't want to be regularly doing a full cluster reboot when something hangs so I'd like to see if I can find what's hung instead.

**Comment by stgraber** [comment]
Does `lxc list --fast` work?

**Comment by stgraber** [comment]
@freeekanayaka

**Comment by chris-sanders** [comment]
Yes that seems to help, several containers are listed as ERROR state.

https://pastebin.ubuntu.com/p/7DBxrFDDZw/

Those containers are all running in fact the last one is the juju controller for the cluster and I can get a status from it so I know it's up.

I notice that using --fast I'm not getting the cluster host that each container is on, is there a way to get that mapping? Maybe the errored hosts are a communication problem to a specific cluster node? I'll check this from each node to see if it's different.

**Comment by chris-sanders** [comment]
Ok, one node doesn't return with --fast and hangs on cluster list as well.

lxd.socket and lxd.service are both running. Socket has been up with no logs since Apr 5.

journalctl for lxd.service doesn't have anything particularly recent but there is a traceback from earlier this month. It's very possible this issue has been active that long since the containers are up and I don't know when the 'list' commanded started hanging.

> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: net/http.(*conn).serve.func1(0xc422794a00)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/net/http/server.go:1726 +0xd0
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: panic(0xed0360, 0xc4201aa740)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/runtime/panic.go:502 +0x229
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/CanonicalLtd/go-dqlite/internal/client.(*Message).bufferForGet(0xc42047c070, 0x434f94)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/CanonicalLtd/go-dqlite/internal/client/message.go:450 +0x96
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/CanonicalLtd/go-dqlite/internal/client.(*Message).getUint64(0xc42047c070, 0x0)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/CanonicalLtd/go-dqlite/internal/client/message.go:379 +0x38
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/CanonicalLtd/go-dqlite/internal/client.DecodeResult(0xc42047c070, 0x11326a0, 0xc4200adf20, 0xc42047c010, 0xc42047c070)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/CanonicalLtd/go-dqlite/internal/client/response.go:177 +0x106
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/CanonicalLtd/go-dqlite.(*Conn).ExecContext(0xc42047c000, 0x11326a0, 0xc4200adf20, 0x1015130, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/CanonicalLtd/go-dqlite/driver.go:303 +0x156
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/CanonicalLtd/go-dqlite.(*Tx).Rollback(0xc42163e188, 0x0, 0x0)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/CanonicalLtd/go-dqlite/driver.go:402 +0xdf
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: database/sql.(*Tx).rollback.func1()
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/database/sql/sql.go:1949 +0x3c
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: database/sql.withLock(0x1129ee0, 0xc4201d1300, 0xc424c13338)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/database/sql/sql.go:3032 +0x63
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: database/sql.(*Tx).rollback(0xc4221d6980, 0xc424c13600, 0x1, 0x1)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/database/sql/sql.go:1948 +0x83
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: database/sql.(*Tx).Rollback(0xc4221d6980, 0xc420a93200, 0xc424c136e0)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/database/sql/sql.go:1963 +0x30
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/lxc/lxd/lxd/db/query.rollback(0xc4221d6980, 0x1124a60, 0xc4201aa300, 0x0, 0xc424c13438)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/db/query/transaction.go:33 +0x2f
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/lxc/lxd/lxd/db/query.Transaction(0xc420298aa0, 0xc424c13460, 0xc424b7e015, 0x10)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/db/query/transaction.go:19 +0xea
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/lxc/lxd/lxd/db.dbQueryRowScan.func1(0xc4239cc558, 0xc424c13560)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/db/db.go:413 +0xac
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/lxc/lxd/lxd/db/query.Retry(0xc424c13540, 0x7f1d6da9fd90, 0x0)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/db/query/retry.go:20 +0xae
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/lxc/lxd/lxd/db.dbQueryRowScan(0xc420298aa0, 0x10694a1, 0xe4, 0xc424c13648, 0x1, 0x1, 0xc424c136e0, 0xa, 0xa, 0x3, ...)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/db/db.go:412 +0xcc
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/lxc/lxd/lxd/db.(*Cluster).ContainerGet(0xc420302360, 0xc421934014, 0xd, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/db/containers.go:657 +0x346
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: main.containerLoadByName(0xc422741dd0, 0xc421934014, 0xd, 0xd, 0x0, 0x0, 0x0)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/container.go:1056 +0x7f
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: main.containerState(0xc4201df8c0, 0xc42098c200, 0x100eff5, 0x3)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/container_state.go:28 +0x17d
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: main.(*Daemon).createCmd.func1(0x112fd60, 0xc420af76c0, 0xc42098c200)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/daemon.go:322 +0x793
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: net/http.HandlerFunc.ServeHTTP(0xc42011eb40, 0x112fd60, 0xc420af76c0, 0xc42098c200)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/net/http/server.go:1947 +0x44
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: github.com/gorilla/mux.(*Router).ServeHTTP(0xc4201e4770, 0x112fd60, 0xc420af76c0, 0xc42098c200)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/gorilla/mux/mux.go:162 +0xed
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: main.(*lxdHttpServer).ServeHTTP(0xc4201ab060, 0x112fd60, 0xc420af76c0, 0xc42020c000)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /build/lxd-j7VLB_/lxd-3.0.3/obj-x86_64-linux-gnu/src/github.com/lxc/lxd/lxd/api.go:79 +0xa4
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: net/http.serverHandler.ServeHTTP(0xc4200d2b60, 0x112fd60, 0xc420af76c0, 0xc42020c000)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/net/http/server.go:2697 +0xbc
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: net/http.(*conn).serve(0xc422794a00, 0x1132620, 0xc4239b2200)
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/net/http/server.go:1830 +0x651
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]: created by net/http.(*Server).Serve
> May 03 13:41:39 juju-89bcd1-10-lxd-1 lxd[921]:         /usr/lib/go-1.10/src/net/http/server.go:2798 +0x27b
> May 21 03:37:59 juju-89bcd1-10-lxd-1 lxd[921]: t=2020-05-21T03:37:59+0000 lvl=warn msg="Raft: Rejecting vote request from 192.168.0.107:8443 since we have a leader: 192.168.0.123:8443"
> May 21 03:38:03 juju-89bcd1-10-lxd-1 lxd[921]: t=2020-05-21T03:38:03+0000 lvl=warn msg="Raft: Heartbeat timeout from \"192.168.0.123:8443\" reached, starting election"
> May 21 04:42:17 juju-89bcd1-10-lxd-1 lxd[921]: t=2020-05-21T04:42:17+0000 lvl=warn msg="Raft: Rejecting vote request from 192.168.0.123:8443 since we have a leader: 192.168.0.107:8443"
> May 21 04:42:21 juju-89bcd1-10-lxd-1 lxd[921]: t=2020-05-21T04:42:21+0000 lvl=warn msg="Raft: Rejecting vote request from 192.168.0.123:8443 since we have a leader: 192.168.0.107:8443"
> 

I can restart the services, or later do a controlled reboot of this node. Before I do though is there anything else I should check to see what's hung?

**Comment by freeekanayaka** [comment]
You can see the container-node mapping with:

```
lxd sql global "SELECT instances.name, nodes.name FROM instances JOIN nodes ON instances.node_id = node.id"
```

**Comment by freeekanayaka** [comment]
Ahh I just noticed that you're running the cluster with the LXD 3.0.x series. The clustering feature was at its early stage in 3.0.x and we did a ton of improvements with 4.0.x. I'd recommend upgrading your cluster to the 4.0.x stable series, it will provide you a better experience and we'll be able to support you more easily.

**Comment by chris-sanders** [comment]
Is that supported/available via packages on bionic? I'm unable to use the snap currently as I have issues mounting rbd volumes in containers on the snap which I've never tracked down.

**Comment by freeekanayaka** [comment]
Only snap.

@stgraber are you aware of any issue around mounting rbd volumes in containers?

**Comment by chris-sanders** [comment]
I went digging, here's the post back when I was setting it up. https://discuss.linuxcontainers.org/t/rbd-map-fails-in-nested-lxd-if-lxd-installed-from-snap/5160

For this issue any more information I should gather or just restart the lxd services and if that doesn't work do a full reboot?

**Comment by chris-sanders** [comment]
Well I can't explain what happened, but here's what has brought everything back.

I tried stopping lxd.socket and lxd.service both hung. I then sshed into each machine and shut them down directly. After that I was able to restart lxd.socket and lxd.service, this restored normal operation for 'lxc list' and the containers all came back up when lxd.service restarted.

I don't see any indication of what happened to cause this, but I don't recall this happening and I've been using it for ~1 year at this point. If this reproduces or if I find a way to reproduce it I'll post back but for now it's working I just don't have a lot of information what happened to that one host.
